### PR TITLE
[WIP] WebGPUTable progress, animometer attributes set externally

### PR DESCRIPTION
**Pull request by ilijapuaca**
Start of `WebGPUTable`/`WebGPUColumn` implementation, `animometer` example revisited

### CODE REVIEW

File: cpp/examples/luma.gl/animometer.cc [hunk 1]
```diff
@@ -44,17 +46,8 @@ layout(std140, set = 0, binding = 0) uniform Constants {
...
-    vec4( 0.1f, -0.1f, 0.0f, 1.0f)
-);
-
-const vec4 colors[3] = vec4[3](
-    vec4(1.0f, 0.0f, 0.0f, 1.0f),
-    vec4(0.0f, 1.0f, 0.0f, 1.0f),
-    vec4(0.0f, 0.0f, 1.0f, 1.0f)
-);
+layout(location = 0) in vec4 positions[3];
+layout(location = 3) in vec4 colors[3];
```
Comment (ibgreen) [inline cpp/examples/luma.gl/animometer.cc:50]: Why location 3?  Add a comment?

File: cpp/modules/deck.gl/core/src/lib/attribute/attribute-manager.cc [hunk 2]
```diff
@@ -46,20 +47,22 @@ void AttributeManager::invalidateAll() {
...
+    auto arrowColumn = descriptor->attributeBuilder(table);
+    // TODO(ilija@unfolded.ai): Ideally we cache the columns and set data into an existing buffer,
+    // but they have fixed size?
+    auto webgpuColumn = std::make_shared<WebGPUColumn>(this->device, arrowColumn);
+    columns.push_back(webgpuColumn);
   }
 
-  auto schema = std::make_shared<arrow::Schema>(fields);
-  return arrow::Table::Make(schema, arrays);
+  return std::make_shared<WebGPUTable>(table->num_rows(), columns);
```
Comment (ibgreen) [inline cpp/modules/deck.gl/core/src/lib/attribute/attribute-manager.cc:67]: Nit for another PR? We should be able to deduce number of rows from columns?

File: cpp/modules/deck.gl/core/src/lib/attribute/attribute-manager.cc [hunk 3]
```diff
@@ -46,20 +47,22 @@ void AttributeManager::invalidateAll() {
...
   for (auto descriptor : this->_descriptors) {
     auto field = std::make_shared<arrow::Field>(descriptor->name, descriptor->type);
     fields.push_back(field);
 
-    auto columnData = descriptor->attributeBuilder(table);
-    arrays.push_back(columnData);
+    auto arrowColumn = descriptor->attributeBuilder(table);
+    // TODO(ilija@unfolded.ai): Ideally we cache the columns and set data into an existing buffer,
+    // but they have fixed size?
+    auto webgpuColumn = std::make_shared<WebGPUColumn>(this->device, arrowColumn);
```
Comment (ibgreen) [inline cpp/modules/deck.gl/core/src/lib/attribute/attribute-manager.cc:63]: I would just call this variable `gpuColumn`...

File: cpp/modules/deck.gl/core/src/lib/attribute/attribute-manager.cc [hunk 4]
```diff
@@ -46,20 +47,22 @@ void AttributeManager::invalidateAll() {
...
+auto AttributeManager::update(const std::shared_ptr<arrow::Table>& table) -> std::shared_ptr<WebGPUTable> {
   // Create an empty output table
   std::vector<std::shared_ptr<arrow::Field>> fields{};
-  std::vector<std::shared_ptr<arrow::Array>> arrays{};
+  std::vector<std::shared_ptr<WebGPUColumn>> columns{};
 
   // Iterate over descriptors and create one column per-descriptor
   for (auto descriptor : this->_descriptors) {
     auto field = std::make_shared<arrow::Field>(descriptor->name, descriptor->type);
     fields.push_back(field);
```
Comment (ibgreen) [inline cpp/modules/deck.gl/core/src/lib/attribute/attribute-manager.cc:58]: What happens to the fields? As you know I think we should still create a schema and store on the GPUTable...

File: cpp/modules/deck.gl/core/src/lib/attribute/attribute-manager.h [hunk 5]
```diff
@@ -67,7 +67,7 @@ class AttributeManager {
   void invalidate(const std::string& attributeName);
   void invalidateAll();
 
-  auto update(const std::shared_ptr<arrow::Table>& table) -> std::shared_ptr<arrow::Table>;
+  auto update(const std::shared_ptr<arrow::Table>& table) -> std::shared_ptr<lumagl::WebGPUTable>;
```
Comment (ibgreen) [inline cpp/modules/deck.gl/core/src/lib/attribute/attribute-manager.h:70]: My sense is that this class should have the GPUTable as a (public) member, and `update()` should return void.

File: cpp/modules/deck.gl/core/test/lib/attribute/attribute-manager-test.cc [hunk 6]
```diff
@@ -77,6 +77,9 @@ TEST_F(AttributeManagerTest, Update) {
   manager->add(std::make_shared<AttributeDescriptor>("attribute-one", arrow::float32(), attributeUpdater));
   manager->add(std::make_shared<AttributeDescriptor>("attribute-two", arrow::float32(), attributeUpdater));
 
+  // TODO(ilija@unfolded.ai): Update tests to use the new WebGPU interface
```
Comment (ibgreen) [inline cpp/modules/deck.gl/core/test/lib/attribute/attribute-manager-test.cc:80]: I would expect `GPUTable` to support `num_rows()` and `num_columns()`. Reading back data is more complicated that can wait a little.

File: cpp/modules/luma.gl/core/src/model.cpp [hunk 7]
```diff
@@ -21,25 +21,44 @@
...
+
+  descriptor.cVertexState.cVertexBuffers[0].arrayStride = sizeof(mathgl::Vector4<float>);
+  descriptor.cVertexState.cVertexBuffers[0].attributeCount = 3;
+  descriptor.cVertexState.cVertexBuffers[0].attributes = &descriptor.cVertexState.cAttributes[0];
+
+  descriptor.cVertexState.cVertexBuffers[1].arrayStride = sizeof(mathgl::Vector4<float>);
+  descriptor.cVertexState.cVertexBuffers[1].attributeCount = 3;
+  descriptor.cVertexState.cVertexBuffers[1].attributes = &descriptor.cVertexState.cAttributes[3];
+
+  for (int i = 0; i < 6; i++) {
```
Comment (ibgreen) [inline cpp/modules/luma.gl/core/src/model.cpp:57]: Hmm, why 6? We have one attribute per buffer...

File: cpp/modules/luma.gl/core/src/model.cpp [hunk 8]
```diff
@@ -50,6 +69,28 @@ Model::Model(std::shared_ptr<wgpu::Device> device, const Model::Options& options
...
-void Model::draw() {
-  // TODO(ilija@unfolded.ai): Implement
+void Model::setAttributes(const std::shared_ptr<WebGPUTable>& table) { this->_attributes = table; }
+
+void Model::draw(wgpu::RenderPassEncoder pass) {
+  if (!this->_attributes) {
+    return;
+  }
+
+  std::vector<BindingInitializationHelper> bindings;
```
Comment (ibgreen) [inline cpp/modules/luma.gl/core/src/model.cpp:79]: Does this have to be done every draw? Can we save this as part of the model and update in setAttributes?

File: cpp/modules/luma.gl/core/src/model.cpp [hunk 9]
```diff
@@ -50,6 +69,28 @@ Model::Model(std::shared_ptr<wgpu::Device> device, const Model::Options& options
...
+  for (auto& attribute : this->_attributes->columns) {
+    BindingInitializationHelper bindingHelper{0, attribute->buffer, 0, attribute->elementSize};
+    bindings.push_back(bindingHelper);
+  }
+
+  wgpu::BindGroup bindGroup =
```
Comment (ibgreen) [inline cpp/modules/luma.gl/core/src/model.cpp:85]: I can't see where this is used?

**Comment by ilijapuaca** [comment]
I pushed my work on top of your commit which was in another branch. There is one issue I'd like to fix, and spent quite some time on, where `animometer` now only renders a single triangle.

This is something that I plan on looking into further, but I suggest we merge this in unless there are major issues with what's in there, as #99 relies on this work already

**Comment by ibgreen** [comment]
Landing. Please address comments in a separate PR.

**Pull request merged**